### PR TITLE
App overview

### PR DESCRIPTION
**Pull request by cicoyle**
This PR intends to:

1. Stub out the `EventPut` functionality to put events (build/deployment/release) into an event table upon upserting the build/deployment/release. Although, this is not actually implemented in boltdb yet, hence why I return `Unimplemented` and the test is in the `Unimplemented` list.
2. Server/State tests for events.
3. The addition of artifact id to some tests that were failing due to the deployment not having the artifact id on it, but this is now necessary for eventPut.
4. The service `UI_ListEvents` func and Stub out the `EventListBundles` functionality to return the list of paginated events. Although, this is not actually implemented in boltdb yet, hence why I return `Unimplemented` and the test is in the `Unimplemented` list.
5. Regenerate the protos with the new event proto updates - you should see 2 new rpcs: `UI_ListEvents` & `INTERNAL_AdditionalMessages` with the corresponding messages + the sorting request needed for pagination.
6. README for what the `pkg/server/ptypes` is intended for.
7. Some typo fixes
9. Server state struct for event

### CODE REVIEW

File: pkg/server/ptypes/README.md
Comment (izaaklauer) [inline pkg/server/ptypes/README.md:1]: Way to be the change you want to see!

File: pkg/server/proto/server.proto
Comment (izaaklauer) [inline pkg/server/proto/server.proto:490]: I think you need an entry in `gateway.yml` to go along with this

File: pkg/server/proto/server.proto
Comment (izaaklauer) [inline pkg/server/proto/server.proto:492]: Could you elaborate a little more on what we expect the frontend (or other api clients) to use this for?

Also likely needs an entry in `gateway.yml`

File: pkg/server/ptypes/build.go
Comment (izaaklauer) [inline pkg/server/ptypes/build.go:13]: Another good catch!

File: pkg/server/ptypes/event.go
Comment (izaaklauer) [inline pkg/server/ptypes/event.go:14]: Is workspace really required?

File: pkg/serverhandler/handlertest/test_service_ui_event.go
Comment (izaaklauer) [inline pkg/serverhandler/handlertest/test_service_ui_event.go:44]: Curious: what is the sleep here accomplishing?

**Review by izaaklauer — CHANGES_REQUESTED** [review]
A few missing entries in gateway.yml, but otherwise looks great!

File: pkg/serverhandler/handlertest/test_service_ui_event.go
Comment (cicoyle) [reply pkg/serverhandler/handlertest/test_service_ui_event.go:44]: My idea for the `time.Sleep` was to have a different Status Start & Complete time for the created resources.

File: pkg/server/ptypes/event.go
Comment (cicoyle) [reply pkg/server/ptypes/event.go:14]: Good catch, probably not.

File: pkg/server/proto/server.proto
Comment (andrew-hashicorp) [reply pkg/server/proto/server.proto:492]: @izaaklauer We don't expect anything to call this, but by defining an RPC that uses `UI.EventBuild / UI.EventDeployment / UI.EventRelease` protos, protoc-gen-swagger picks them up when generating the swagger. The swagger is then used when the frontend by openapi-generator and will generate TypeScript types for `UI.EventBuild / UI.EventDeployment / UI.EventRelease`.

We can then use these on the frontend when unmarshaling the untyped EventBundle.EventData field.

File: pkg/serverhandler/handlertest/test_service_ui_event.go
Comment (andrew-hashicorp) [reply pkg/serverhandler/handlertest/test_service_ui_event.go:44]: My understanding is that we emit 2 events for the same operation, one with status.startTime filled in, and one with both status.startTime + status.completeTime filled in. Is that correct?

At the state layer, we then upsert events so the same event gets updated with a completeTime when it's complete?

On the frontend then, depending on when we call UI_ListEvents, we could get an event with either a startTime but no completeTime or both startTime and completeTime, but we would always only get one event per operation?

File: pkg/server/proto/server.proto
Comment (cicoyle) [reply pkg/server/proto/server.proto:490]: I just added that in.

File: pkg/server/proto/server.proto
Comment (cicoyle) [reply pkg/server/proto/server.proto]: Thanks @andrew-hashicorp for elaborating. I updated the description comment line above to read: `  // List possible event data types.` per my discussion with Andrew. LMK if yall feel like that could be better explained.

Comment (izaaklauer) [reply pkg/server/proto/server.proto]: Ah interesting. I suppose it makes sense that the types wouldn't make it all the way through to the final js client if it isn't used by an rpc here.

Can we add that as a comment on this RPC then, and rename it to something that makes it clear potential api callers can ignore this? Something like

**Review by izaaklauer — APPROVED** [review]
This looks good overall! Small request for comments and updating naming for rpc endpoints we don't ever expect to call, but that shouldn't block merging this.

File: pkg/serverhandler/handlertest/test_service_ui_event.go
Comment (cicoyle) [reply pkg/serverhandler/handlertest/test_service_ui_event.go:45]: I believe that is correct. As UI_ListEvents calls the EventListBundles function which will list the events from the db. An event will have a completed time in the status field only after its completed and the second Upsert<Operation> call is made from the runner.

Each Upsert<Build/Deploy/Release> gets called upon the correlative cli commands, that puts the operation in the db, then the runner will call it again to update that record to put in the completed time once its done. EventPut operates in the same way as it gets called in the Upsert<Operation>. This will put an event the first time it is 'started', then once the runner updates the operation record the second time it calls Upsert<Operation> it will update the correlative event record with a new EventTimestamp and eventData containing that completed time.

File: pkg/server/proto/server.proto
Comment (andrew-hashicorp) [inline pkg/server/proto/server.proto:684]:
```suggestion
    EventType event_type  = 2;
    bytes event_data = 3;
```
This was a mistake in the RFC (since corrected) - Should be in snake_case to follow the convention established in the rest of `server.proto`.

File: pkg/server/proto/server.proto
Comment (andrew-hashicorp) [inline pkg/server/proto/server.proto:733]:
```suggestion
    UI.EventBuild event_build_schema = 1;
    UI.EventDeployment event_deployment_schema = 2;
    UI.EventRelease event_release_schema = 3;
```
Same as above - should be snake_case to follow convention.

**Review by andrew-hashicorp — CHANGES_REQUESTED** [review]

**Review by andrew-hashicorp — CHANGES_REQUESTED** [review]

File: pkg/server/proto/gateway.yml
Comment (cicoyle) [reply pkg/server/proto/gateway.yml]: oh really? I updated the endpoint accordingly.

**Comment by cicoyle** [comment]
fixing merge conflicts now.

File: pkg/server/proto/gateway.yml
Comment (izaaklauer) [inline pkg/server/proto/gateway.yml:559]: can we modify this name to also be internal additional messages instead of event schemas?

Comment (cicoyle) [reply pkg/server/proto/gateway.yml:559]: ahh, yes I forgot about that. yeah I'll update that

Comment (izaaklauer) [reply pkg/server/proto/gateway.yml:559]: After that, merge away!

Comment (cicoyle) [reply pkg/server/proto/gateway.yml]: done

**Pull request merged**
